In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import pandas as pd
from sklearn.metrics import mean_absolute_error

In [2]:
df = pd.read_csv("energydata_complete.csv",encoding='latin-1')

In [3]:
df

date  Appliances  lights         T1       RH_1  \
0      2016-01-11 17:00:00          60      30  19.890000  47.596667   
1      2016-01-11 17:10:00          60      30  19.890000  46.693333   
2      2016-01-11 17:20:00          50      30  19.890000  46.300000   
3      2016-01-11 17:30:00          50      40  19.890000  46.066667   
4      2016-01-11 17:40:00          60      40  19.890000  46.333333   
...                    ...         ...     ...        ...        ...   
19730  2016-05-27 17:20:00         100       0  25.566667  46.560000   
19731  2016-05-27 17:30:00          90       0  25.500000  46.500000   
19732  2016-05-27 17:40:00         270      10  25.500000  46.596667   
19733  2016-05-27 17:50:00         420      10  25.500000  46.990000   
19734  2016-05-27 18:00:00         430      10  25.500000  46.600000   

              T2       RH_2         T3       RH_3         T4  ...         T9  \
0      19.200000  44.790000  19.790000  44.730000  19.000000  ...  17.033333   
1      19.200000  44.722500  19.790000  44.790000  19.000000  ...  17.066667   
2      19.200000  44.626667  19.790000  44.933333  18.926667  ...  17.000000   
3      19.200000  44.590000  19.790000  45.000000  18.890000  ...  17.000000   
4      19.200000  44.530000  19.790000  45.000000  18.890000  ...  17.000000   
...          ...        ...        ...        ...        ...  ...        ...   
19730  25.890000  42.025714  27.200000  41.163333  24.700000  ...  23.200000   
19731  25.754000  42.080000  27.133333  41.223333  24.700000  ...  23.200000   
19732  25.628571  42.768571  27.050000  41.690000  24.700000  ...  23.200000   
19733  25.414000  43.036000  26.890000  41.290000  24.700000  ...  23.200000   
19734  25.264286  42.971429  26.823333  41.156667  24.700000  ...  23.200000   

          RH_9      T_out  Press_mm_hg     RH_out  Windspeed  Visibility  \
0      45.5300   6.600000        733.5  92.000000   7.000000   63.000000   
1      45.5600   6.483333        733.6  92.000000   6.666667   59.166667   
2      45.5000   6.366667        733.7  92.000000   6.333333   55.333333   
3      45.4000   6.250000        733.8  92.000000   6.000000   51.500000   
4      45.4000   6.133333        733.9  92.000000   5.666667   47.666667   
...        ...        ...          ...        ...        ...         ...   
19730  46.7900  22.733333        755.2  55.666667   3.333333   23.666667   
19731  46.7900  22.600000        755.2  56.000000   3.500000   24.500000   
19732  46.7900  22.466667        755.2  56.333333   3.666667   25.333333   
19733  46.8175  22.333333        755.2  56.666667   3.833333   26.166667   
19734  46.8450  22.200000        755.2  57.000000   4.000000   27.000000   

       Tdewpoint        rv1        rv2  
0       5.300000  13.275433  13.275433  
1       5.200000  18.606195  18.606195  
2       5.100000  28.642668  28.642668  
3       5.000000  45.410389  45.410389  
4       4.900000  10.084097  10.084097  
...          ...        ...        ...  
19730  13.333333  43.096812  43.096812  
19731  13.300000  49.282940  49.282940  
19732  13.266667  29.199117  29.199117  
19733  13.233333   6.322784   6.322784  
19734  13.200000  34.118851  34.118851  

[19735 rows x 29 columns]

In [4]:
df.shape

(19735, 29)

In [5]:
df.columns

Index(['date', 'Appliances', 'lights', 'T1', 'RH_1', 'T2', 'RH_2', 'T3',
       'RH_3', 'T4', 'RH_4', 'T5', 'RH_5', 'T6', 'RH_6', 'T7', 'RH_7', 'T8',
       'RH_8', 'T9', 'RH_9', 'T_out', 'Press_mm_hg', 'RH_out', 'Windspeed',
       'Visibility', 'Tdewpoint', 'rv1', 'rv2'],
      dtype='object')

## Fitting a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). 

In [6]:
X = df[['T2']]
y = df['T6']

# Split the data into a training set and a testing set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [7]:
# Create and fit the linear regression model
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [8]:
# Make predictions on the test set
y_pred = model.predict(X_test)

## Calculating the Root Mean Squared error in three D.P

In [9]:
# Calculate RMSE
rmse = sqrt(mean_squared_error(y_test, y_pred))

print(f'Root Mean Squared Error (RMSE): {rmse:.3f}')

Root Mean Squared Error (RMSE): 3.633


##  Running a multiple linear regression using the training set and calculating the Mean Absolute Error (in three decimal places) for the  training set

In [9]:
from sklearn.preprocessing import MinMaxScaler

In [10]:
# Remove columns "date" and "lights"
df = df.drop(['date', 'lights'], axis=1)

In [11]:
target = df["Appliances"]
features = df.drop("Appliances", axis=1)

In [12]:
# Perform a 70-30 train-test split with a random state of 42
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.3, random_state=42)

In [13]:
# Initialize and fit the MinMaxScaler
scaler = MinMaxScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [14]:
# Create and fit a multiple linear regression model
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [15]:
# Make predictions on the training set
y_train_pred = model.predict(X_train)

In [16]:
# Calculate the Mean Absolute Error for the training set
mae = mean_absolute_error(y_train, y_train_pred)

In [17]:
print(f"Mean Absolute Error for the training set: {mae:.3f}")

Mean Absolute Error for the training set: 53.742


## Calculating the Root Mean Squared Error (in three decimal places) for the training set

In [18]:
import math

In [19]:
# Make predictions on the training set
y_train_pred = model.predict(X_train)

# Calculate the Mean Squared Error (MSE)
mse = mean_squared_error(y_train, y_train_pred)

# Calculate the RMSE
rmse = math.sqrt(mse)

print(f"Root Mean Squared Error for the training set: {rmse:.3f}")

Root Mean Squared Error for the training set: 95.216


## Calculating the Mean Absolute Error (in three decimal places) for test set

In [20]:
# Make predictions on the test set
y_test_pred = model.predict(X_test)

# Calculate the Mean Absolute Error (MAE)
mae_test = mean_absolute_error(y_test, y_test_pred)

print(f"Mean Absolute Error for the test set: {mae_test:.3f}")

Mean Absolute Error for the test set: 53.643


##  Calculating the Root Mean Squared Error (in three decimal places) for test set

In [21]:
# Make predictions on the test set
y_test_pred = model.predict(X_test)

# Calculate the Mean Squared Error (MSE)
mse_test = mean_squared_error(y_test, y_test_pred)

# Calculate the RMSE
rmse_test = math.sqrt(mse_test)

print(f"Root Mean Squared Error for the test set: {rmse_test:.3f}")

Root Mean Squared Error for the test set: 93.640


## Training a ridge regression model with default parameters and finding out if there is any change to the root mean squared error (RMSE) when evaluated on the test set

In [29]:
from sklearn.linear_model import Ridge

# Create and train a Ridge regression model with default parameters
ridge_model = Ridge()
ridge_model.fit(X_train, y_train)

# Make predictions on the test set
y_test_pred = ridge_model.predict(X_test)

# Calculate the Mean Squared Error (MSE)
mse_test = mean_squared_error(y_test, y_test_pred)

# Calculate the RMSE
rmse_test = math.sqrt(mse_test)

print(f"Root Mean Squared Error for the test set with Ridge regression: {rmse_test:.3f}")

Root Mean Squared Error for the test set with Ridge regression: 93.709


Conclusion: There isn't any significant change as the RMSE for the test set with Ridge regression is 93.709 which is not far from the Root Mean Squared Error for the test set which is 93.640

## Training a lasso regression model with default value and obtaining the new feature weights with it to know many of the features have non-zero feature weights

In [22]:
from sklearn.linear_model import Lasso

In [23]:
# Create and train a Lasso regression model with default parameters
lasso_model = Lasso()
lasso_model.fit(X_train, y_train)

# Get the feature weights (coefficients)
feature_weights = lasso_model.coef_

# Count the number of features with non-zero weights
non_zero_features = sum(feature_weights != 0)

print(f"Number of features with non-zero weights: {non_zero_features}")

Number of features with non-zero weights: 4


## Calculating  the new RMSE with the Lasso Regression on the test set

In [24]:
# Make predictions on the test set using the trained Lasso model
y_test_pred_lasso = lasso_model.predict(X_test)

# Calculate the Mean Squared Error (MSE)
mse_test_lasso = mean_squared_error(y_test, y_test_pred_lasso)

# Calculate the RMSE with Lasso Regression
rmse_test_lasso = math.sqrt(mse_test_lasso)

print(f"Root Mean Squared Error for the test set with Lasso Regression: {rmse_test_lasso:.3f}")

Root Mean Squared Error for the test set with Lasso Regression: 99.424
